### PR TITLE
fix(obj): fix layout bugs and optimise layout/style performance

### PR DESCRIPTION
**Pull request by AndyEveritt**
This PR will be a collection of optimisations to improve lvgl's performance.

Particularly I am focusing on program start time for complex UI's where a large chunk of the UI is instantiated at program start, but the optimisations should still improve performance after initialisation by preventing unnecessary computations

### CODE REVIEW

**Comment by AndyEveritt** [comment]
# Change 1 https://github.com/lvgl/lvgl/pull/9731/commits/c5c52bb739d346effbe17df53492c94c5a10a73b

(I generated the summary with AI but it has been reviewed)

1. `lv_obj_mark_layout_as_dirty` - propagate subtree-dirty flag up the parent chain
Previously this only set scr_layout_inv on the screen object. Now it walks from the dirty object up through each ancestor, setting scr_layout_inv on each, with an early break if an ancestor is already marked. This enables subtree pruning during layout updates.

2. `layout_update_core` - skip clean subtrees
Added an early-exit at the top: if `layout_inv`, `readjust_scroll_after_layout`, and `scr_layout_inv` are all clear, the entire subtree is skipped. Also inlined the child count access (`obj->spec_attr->child_cnt`) to avoid the `lv_obj_get_child_count` function call and its assert overhead in the hot recursive path.

After processing, the `scr_layout_inv` flag is recomputed by scanning direct children — it's only cleared if no child was re-dirtied during processing. This handles the case where processing one child triggers `lv_obj_mark_layout_as_dirty` on a previously-processed sibling.

3. `lv_obj_update_layout` - remove redundant screen flag clearing
The outer while loop no longer clears `scr->scr_layout_inv = 0` before calling `layout_update_core`, since `layout_update_core` now manages the flag correctly via the recomputation step.

Performance impact
Steady-state updates (e.g., a single label or value change): only the dirty subtree path is visited instead of the entire widget tree. For a tree of ~200 objects where only one leaf changes, this reduces work from O(200) to O(depth + siblings-at-each-level).
Initial layout (everything dirty): negligible overhead — the early exit doesn't trigger, and the child-flag recomputation is a cheap linear scan of direct children.
No additional memory: reuses the existing scr_layout_inv bit, which was previously unused on non-screen objects.

image

(with log axis scale)
image


Red trace is the before, yellow is after. Notice that the number of calls to `layout_update_core` went from 1,183,805 to 37,423 so even though the mean time is a little longer, the reduced overhead gives a noticeable net improvement taking the time spent in layout_update_core from 3.34 seconds on program start, down to 0.9 seconds

**Comment by AndyEveritt** [comment]
# Change 2
`lv_obj_enable_style_refresh` is now per object instead of global with a new `lv_enable_style_refresh` function added that acts as a global override.

lvgl only uses the per object function leaving the global function for the programmer.

When initialising a UI with a lot of dynamically sized objects (LV_SIZE_CONTENT or LV_PCT), it is not necessary to refresh the styles until all the objects have been initialised.

There was no way to completely prevent lvgl from refreshing styles since the previous global used to disable refreshing styles was toggled every time an object was created. This meant even if you disabled it before creating the UI, after the first object had been created it would be enabled again so you lose the benefit.

Changing to a per object disable doesn't add any noticeable difference to the performance but it does allow for the programmer to use the global disable without worrying about lvgl resetting it.

In initialising my program, the UI creation time is ~30% quicker when disabling style refreshing until after all the objects had been created and styles applied.

image

**Comment by AndyEveritt** [comment]
Ok, so a lot has changed here. There have been some really convoluted edge cases that break the layout system. I have attempted to fix all of them that I am aware of. You will probably ask for unit tests to validate but the scenarios to trigger the layout issues required such a complex setup of objects I wasn't realistically able to create a unit test for most of the scenarios I have fixed. All the issues did occur in a end UI program and are now fixed.

A none exhaustive list of layout edge cases that are now fixed:
1. lv_textarea could cause an infinite loop because `lv_textarea_scroll_to_cusor_pos` would retrigger the `LV_EVENT_STYLE_CHANGED` event if no animation was used (which then calls `lv_textarea_scroll_to_cusor_pos`)
2. lv_label now correctly calculates it's size when min width is LV_SIZE_CONTENT (before it wasn't guaranteed)
3. There was possible infinite loop when sizing lv_textarea because the min and standard label widths were the wrong way round
4. flex and grid layout now both calculate their min size in both directions to prevent an issue where they end up fixed too large and can't reduce in size because the LV_SIZE_CONTENT calculation used the positions of the items and the items used the size of the container
    - This did result in the grid LV_SIZE_CONTENT sometimes being 1 pixel too big per track because of how grid handles rounding
5. Ensure nested layouts are correctly invalidated by invalidating the full parent chain
    - This adds minimal overhead to the processing but is required to handle a frustratingly obscure layout issue I was having

Some optimisations:
1. Made most of the layout related calculations that could use `const lv_obj_t*` use it to prevent accidental modification to objects at the wrong time
2. Only process the layout for none hidden objects


This is a big change but I am done with this PR for now. @AndreCostaaa would you be able to review the changes? I imagine given the extent of the files touched it will quickly have some merge conflicts

File: src/core/lv_obj_pos.c [hunk 1]
```diff
@@ -181,70 +181,120 @@ bool lv_obj_refr_size(lv_obj_t * obj)
...
             w_style = lv_obj_get_style_max_width(obj, LV_PART_MAIN);
         }
         else {
             w_style = lv_obj_get_style_width(obj, LV_PART_MAIN);
         }
-        obj->w_ignore_size =
-            (LV_COORD_IS_PCT(w_style) && parent->w_layout == 0 && lv_obj_get_style_width(parent, 0) == LV_SIZE_CONTENT);
+        w = LV_CLAMP(minw, w, maxw);
     }
+    obj->w_ignore_size = LV_COORD_IS_PCT(w_style);
```
[the remark this answers is not in the document]
Comment (AndyEveritt) [reply src/core/lv_obj_pos.c:250]: This is ok since the only time w_ignore_size is checked is when doing a LV_SIZE_CONTENT calculation

**Comment by AndyEveritt** [comment]
v9.5.0 has been merged in

**Comment by AndyEveritt** [comment]
It is worth noting that the layout issues were not introduced by the optimisations but were present in mainline lvgl. The fixes ended up being related to the performance optimisations which is why I have done 1 PR.

I have also found 1 more layout issue that was in mainline lvgl which I think I have a fix for but I'm tested currently

File: src/core/lv_obj_style.c [hunk 2]
```diff
@@ -882,7 +878,7 @@ static void report_style_change_core(void * style, lv_obj_t * obj)
     for(i = 0; i < obj->style_cnt; i++) {
         if(style == NULL || obj->styles[i].style == style) {
             full_cache_refresh(obj, lv_obj_style_get_selector_part(obj->styles[i].selector));
-            lv_obj_refresh_style(obj, LV_PART_ANY, LV_STYLE_PROP_ANY);
+            refresh_obj_style_core(obj, LV_PART_ANY, LV_STYLE_PROP_ANY);
```
[the remark this answers is not in the document]
Comment (AndyEveritt) [reply src/core/lv_obj_style.c:881]: Objects are recursed later in the function

**Comment by AndreCostaaa** [comment]
Hey @AndyEveritt 
 
What's the status here?
